### ISSUE TITLE
Auth without using FlowFuse

### ISSUE THREAD

**Issue by fullmetal-fred** [FlowFuse/node-red-dashboard#638]
### Description

Hi team FlowFuse! 

Congrats on the recent release and on the awesome demo of multi-user dashboards!

In light of that I have a request: could we allow an auth workflow for dashboard on standard OSS Node-Red? I love using FlowFuse, don't get me wrong, but I teach at tradeshows where I don't have a reliable internet connection. I could perhaps self-host FlowFuse and run it on prem, and I'm open to that, but I'd also just love to be able to use a simple local auth flow. 

This could be basic auth...storing the username & password as a cookie and passing them via headers on every request. This could also be something more like a session token / JWT...where the user authenticates, a flow validates the auth, then sends a JWT as a cookie back to the user...then every subsequent request from that client would include additional session & user id information in the ._client object. 

I'm not particularly picky how...but I do need a unique way to authenticate & keep track of a unique user. It does seem that this could be done with a dashboard-2 plugin...and I'm also open to working on that if we feel that's the right approach. But I wanted to ask first if this should be native functionality in the opinion of the maintainers. 

Then again, you tell me: if I can do this with self-hosted FlowFuse, I'll look to do that for now. 

Thanks!!

### Have you provided an initial effort estimate for this issue?

I am no FlowFuse team member

**Comment by wazzaarr** [comment]
I agree, while I do want to support flowfuse, pay walling out this feature in particular doesn't seem right. 
It's so highly requested in Node-RED. 
Also making it only available at $125 a month, skipping the $15 a month doesn't push Node-RED users into Flowfuse.

**Comment by joepavitt** [comment]
> In light of that I have a request: could we allow an auth workflow for dashboard on standard OSS Node-Red? I love using FlowFuse, don't get me wrong, but I teach at tradeshows where I don't have a reliable internet connection. I could perhaps self-host FlowFuse and run it on prem, and I'm open to that, but I'd also just love to be able to use a simple local auth flow.

@fullmetal-fred When you're teaching at the tradeshows, does that mean you're allowing access to all of your users/students to your own Dashboard?

**Comment by joepavitt** [comment]
> While I do want to support flowfuse, pay walling out this feature in particular doesn't seem right.
It's so highly requested in Node-RED.

@wazzaarr I'd be very interested to know more about your use case. Our current decision is based on a case that those wanting authenticated multi-user authenticated Dashboards would be a professional/industry/enterprise requirement, running with a team of users wanting access.

**Comment by wazzaarr** [comment]
Node-RED has become so much more, due to your work on dashboard 2. Where it was really more applicable to IOT in the past, is now really opened up to being one of the best low-code web app development platforms. Fundamentally, multi user is a requirement of web app development.

**Comment by wazzaarr** [comment]
That's great info.
I've actually started using WeWeb as the front end. It works really well.

**Comment by joepavitt** [comment]
> I've actually started using WeWeb as the front end

Is that in-place of Dashboard 2.0, or supplementary to?

**Comment by wazzaarr** [comment]
In place of, Joe.
Absolutely need multi user functionality.
Node-RED is a very powerful middleware and had a huge place in the low code
movement.

**Comment by uschmelmer** [comment]
I would say multi-user or multi-dashboard capability is an absolute need, also for the selfhosted users.
Without these features i can't find any advantage of DB 2.0 compared to the original version.

Kind Regards

**Comment by mman** [comment]
> Without these features i can't find any advantage of DB 2.0 compared to the original version.

Like for example that DB 2.0 is actively maintained, does not use deprecated, and unsupported, and possibly insecure Angular 1. 😉

**Comment by joepavitt** [comment]
> I would say multi-user or multi-dashboard capability is an absolute need, also for the selfhosted users.

In which case, FlowFuse is available to support this use-case.

**Comment by uschmelmer** [comment]
> > I would say multi-user or multi-dashboard capability is an absolute need, also for the selfhosted users.
> 
> In which case, FlowFuse is available to support this use-case.

We se need for Industral Control, for example you have some machine/productionline controlpanl with three levels of access:
1. Normal Worker: Monitoring Machine Progress and simple controls like start/stop, tool change
2. Machine Operator: Changing machine settings,  error clearing
3. Service: Chaning lowlevel settings, detailed error log

Same for building automation, example hotel rooms:
1. Guest Accounts: Change Light, Sound Volume
2. Advanced settings: like HVAC, Timers

Best Regards

**Comment by joepavitt** [comment]
Thanks for the details @uschmelmer - that's exactly the kind of use cases we have in mind. If you're interested in [FlowFuse](https://flowfuse.com/) as a solution here, I'm happy to put you in touch with someone that can provide more details.

**Comment by joepavitt** [comment]
@cometta we have a few available - listed at https://dashboard.flowfuse.com/user/multi-tenancy.html#understanding-client-data

**Comment by carlbomsdata** [comment]
I have something.

Inside settings.js, place this somewhere in module.exports:
```
    //for dashboard protection
    dashboard: {
        path: "dashboard",
        middleware: require("./custom-login.js")
    },
```

Create a file called custom-login.js (in same folder as settings.js) and paste this:
```
// https://github.com/htmltiger/node-red-custom-login/

const cookie = require('cookie');
const crypto = require('crypto');

const nodeusername = "lemling";
const nodepassword = "2";
const myhash = crypto.createHash('sha512').update("#" + nodeusername + "-" + nodepassword, 'utf8').digest('hex');
const ipban = {};

function customLogin(req, res, next) {
    let logy = '';
    let code = 200;

    if (req.headers.cookie) {
        logy = cookie.parse(req.headers.cookie)["log"] || '';
    }

    if (logy) {
        const ip = req.connection.remoteAddress;
        if (!(ip in ipban)) {
            ipban[ip] = { lastf: 0, attempts: 0 };
        }

        if (ipban[ip].attempts > 0 && Date.now() - ipban[ip].lastf > (15 * 60000)) {
            ipban[ip].attempts = 0;
        }

        if (logy === myhash) {
            res.cookie("user", nodeusername, { expires: new Date(Date.now() + (14 * 864e5)), path: '/' });
            req.user = nodeusername;
            return next();
        }

        const rehashed = crypto.createHash('sha512').update('#' + logy, 'utf8').digest('hex');
        if (ipban[ip].attempts < 3 && rehashed === myhash) {
            alertme("PASSED: " + ip);
            res.cookie("log", myhash, { expires: new Date(Date.now() + (14 * 864e5)), path: '/' });
            res.cookie("user", nodeusername, { expires: new Date(Date.now() + (14 * 864e5)), path: '/' });
            delete ipban[ip];
            req.user = nodeusername;
            return res.redirect(req.originalUrl);
        } else {
            alertme("FAILED: " + ip + " " + logy, req.headers);
            res.clearCookie("log");
            ipban[ip].lastf = Date.now();
            ipban[ip].attempts++;
            logy = "Login Failed";
            code = 401;
        }
    }

    res.writeHeader(code, { "Content-Type": "text/html" });
    res.write('<!DOCTYPE html><html><head><meta name="viewport" content="width=device-width, initial-scale=1" /><style>body{background:#3498DB}');
    res.write('form{margin:20px auto;width:300px;background:#fff;border-radius:5px;padding:20px;text-align:center;color:red}input{padding:12px 0;');
    res.write('margin-bottom:10px;border-radius:3px;border:1px solid transparent;text-align:center;width:100%;font-size:16px}.f{background:#ECF0F1}');
    res.write('.f:focus{border:1px solid #3498DB}.b{background:#3498DB;color:#fff}.b:hover,.b:active{background:#1F78B4}</style></head><body><form ');
    res.write('method="POST" onsubmit="document.cookie=\'log=\'+this.querySelector(\'#u\').value+\'-\'+this.querySelector(\'#p\').value+\'; ');
    res.write('path=/\';return true;"><input type="text" placeholder="Username" id="u" name="username" class="f"><input type="password" ');
    res.write('placeholder="Password" id="p" name="password" class="f"><input type="submit" value="Login" class="b">' + logy + '</form></body></html>');
    return res.end();
}

function alertme(msg, headers) {
    if (headers) {
        headers = JSON.stringify(headers, null, 2);
    } else {
        headers = '';
    }
    msg = "nodelogin: " + msg;
    console.warn("[warn] " + msg);
}

module.exports = customLogin;

```

Restart your nodered instance

Then inside Nodered editor create a flow and add the following (connect them in order below):
* ui-event node (UI: My Dashboard /dashboard)
* ui-template node (Type: widget, UI: My Dashboard) and add the following Template script:
```
<template>
  <div></div> <!-- nothing visual needed -->
</template>

<script>
  export default {
    mounted() {
      const cookie = document.cookie
        .split('; ')
        .find(row => row.startsWith('user='));

      if (cookie) {
        const user = cookie.split('=')[1];
        this.send({ payload: { user: user } });
      }
    }
  }
</script>
```
* change node (Set: msg.payload, To the value: msg.payload.user)
* text node (Group: choose any page and group)

Then add a logout button (not connected to the nodes above):
* template node (Type: Widget, UI: MyDashboard) and add the following Template script:
```
<template>
  <v-btn @click="logout" color="error">Logout</v-btn>
</template>

<script>
export default {
  methods: {
    logout() {
      // Clear cookies by expiring them
      document.cookie = "user=; expires=Thu, 01 Jan 1970 00:00:00 UTC; path=/;";
      document.cookie = "log=; expires=Thu, 01 Jan 1970 00:00:00 UTC; path=/;";
      // Redirect to login page
      window.location.href = "/dashboard";
    }
  }
}
</script>

```
